### PR TITLE
Added ceph-installer provisioning wrapper

### PR DESCRIPTION
**Pull request by shtripat**
The below sample code explains the usage of the this wrapper

```
// # Create the provisioner instance
manager = ProvisioningManager(
    "CephInstallerPlugin"
)
prov_obj = manager.get_plugin()
prov_obj.set_provisioner_node(
    "<provisioner node fqdn>"
)

// # Execute mon installation. Run this for all the mons
task_id = prov_obj.install_mon(["<mon node fqdn 1>", "<mon node fqdn 2>"])
ret_val = task_status(task_id, prov_obj)
if ret_val is False:
    return

// # Execute osd installation. Run this for all the osds
task_id = prov_obj.install_osd([
    "<osd node fqdn 1>",
    "<osd node fqdn 2>"
])
ret_val = task_status(task_id, prov_obj)
if ret_val is False:
    return

// # Execute mon configuration. Run this for all the mons
// # Pass None for first mon case and second mon onwards
// # pass a list of earlier configured mons as
// # [{"host": "mon0.host", "interface": "eth0"}, ....]
task_id = prov_obj.configure_mon(
    "<mon node fqdn>",
    "d5f2afab-f188-45e8-9af6-8c587dca0977",
    "ceph",
    "10.70.40.0/22",
    "10.70.40.0/22",
    None
)
ret_val = task_status(task_id, prov_obj)
if ret_val is False:
    return

// # Execute osd configuration. Rund this for all the osds
// # Pass mon details as last argument to the function as
// # [{"host": "mon0.host", "interface": "eth0"}, {"host": "mon1.host", "address": "10.0.0.1"}]
// # regarding journals, pass devices list as ["/dev/vdb", "/dev/vdc"] if collocated
// # journals required.
// # To have journals on different disk, pass the devices details as
// # {"device1": "journal1", "device2": "journal2", ...}
task_id = prov_obj.configure_osd(
    "<osd node fqdn>",
    ["/dev/vdb", "/dev/vdc"],
    "d5f2afab-f188-45e8-9af6-8c587dca0977",
    "ceph",
    1024,
    "10.70.40.0/22",
    "10.70.40.0/22",
    [{"host": "<mon node fqdn>", "interface": "eth0"}]
)
ret_val = task_status(task_id, prov_obj)
if ret_val is False:
    return

// # Function which waits for 15 mins for a job and reports accordingly
def task_status(task_id, prov_obj):
    for count in range(90):
        gevent.sleep(10)
        status = prov_obj.task_status(task_id)
        if status['ended'] != "" and status['succeeded'] is True:
            return True
    LOG.error("Time out the task")
    return False
```

Refer http://docs.ceph.com/ceph-installer/docs/ for more details on parameter details

Note: While testing the changes its figured out that if ceph-installer installed on one of the mons within cluster, the ceph mon package installation for this mon node fails using ceph-installer call. Because of this the ceph-installer needs to be outside the cluster and node-agent also should be installed on this provisioning node. The cluster provisioning job should be specifically targeted to this provisioning node and things works as required.

### CODE REVIEW

File: tendrl/node_agent/provisioner/ceph/plugins/ceph_installer.py [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+    def __init__(
+        self,
+    ):
+        self.http = urllib3.PoolManager()
+
+    def set_provisioner_node(self, provisioner_node):
+        self.provisioner_node = provisioner_node
+
+    def install_mon(self, mons):
+        url = 'http://localhost:%s/api/mon/install' % \
```
Comment (r0h4n) [inline tendrl/node_agent/provisioner/ceph/plugins/ceph_installer.py]: please provide sample of "mons"

File: tendrl/node_agent/provisioner/ceph/plugins/ceph_installer.py [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+            url,
+            body=json.dumps(data),
+            headers={'Content-Type': 'application/json'})
+        if resp.status == 200:
+            try:
+                res_data = json.loads(resp.data.decode('utf-8'))
+            except (TypeError, ValueError, UnicodeError) as e:
+                raise Exception(
+                    'Server response was not valid JSON: %r' % e)
+            return res_data['identifier']
```
Comment (r0h4n) [inline tendrl/node_agent/provisioner/ceph/plugins/ceph_installer.py]: what is 'identifier'?

File: tendrl/node_agent/provisioner/ceph/plugins/ceph_installer.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+        else:
+            return None
+
+    def install_osd(self, osds):
```
Comment (r0h4n) [inline tendrl/node_agent/provisioner/ceph/plugins/ceph_installer.py]: sample of "osds" please

File: tendrl/node_agent/provisioner/ceph/plugins/ceph_installer.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+            self._CEPH_INSTALLER_API_PORT
+        data = {
+            "calamari": False,
+            "conf": {"global": {"auth supported": "cephx"}},
+            "host": host,
+            "interface": "eth0",
+            "fsid": cluster_id,
+            "monitor_secret": "AQA7P8dWAAAAABAAH/tbiZQn/40Z8pr959UmEA==",
+            "cluster_name": cluster_name,
+            "cluster_network": cluster_network,
```
Comment (r0h4n) [inline tendrl/node_agent/provisioner/ceph/plugins/ceph_installer.py]: please put samples here

File: tendrl/node_agent/provisioner/ceph/plugins/ceph_installer.py [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+        resp = self.http.request(
+            self._MGET,
+            url)
+        if resp.status == 200:
+            try:
+                res_data = json.loads(resp.data.decode('utf-8'))
+            except (TypeError, ValueError, UnicodeError) as e:
+                raise Exception(
+                    'Server response was not valid JSON: %r' % e)
+            return res_data
```
Comment (r0h4n) [inline tendrl/node_agent/provisioner/ceph/plugins/ceph_installer.py]: please provide sample for res_data

File: tendrl/node_agent/provisioner/ceph/manager.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+        try:
+            path = os.path.dirname(os.path.abspath(__file__)) + '/plugins'
+            pkg = 'tendrl.node_agent.provisioner.ceph.plugins'
+            for py in [f[:-3] for f in os.listdir(path)
+                       if f.endswith('.py') and f != '__init__.py']:
+                plugin_name = '.'.join([pkg, py])
+                mod = importlib.import_module(plugin_name)
+                clsmembers = inspect.getmembers(mod, inspect.isclass)
+                for name, cls in clsmembers:
+                    exec("from %s import %s" % (plugin_name, name))
```
Comment (shtripat) [inline tendrl/node_agent/provisioner/ceph/manager.py]: ok. will try that.

File: tendrl/node_agent/provisioner/ceph/plugins/ceph_installer.py (hunk 1 above)
Comment (shtripat) [inline tendrl/node_agent/provisioner/ceph/plugins/ceph_installer.py]: Done

File: tendrl/node_agent/provisioner/ceph/plugins/ceph_installer.py (hunk 2 above)
Comment (shtripat) [inline tendrl/node_agent/provisioner/ceph/plugins/ceph_installer.py]: `identifier` here is job id for http request. We need to keep tracking the same and find if its completed.

File: tendrl/node_agent/provisioner/ceph/plugins/ceph_installer.py (hunk 3 above)
Comment (shtripat) [inline tendrl/node_agent/provisioner/ceph/plugins/ceph_installer.py]: Done

File: tendrl/node_agent/provisioner/ceph/plugins/ceph_installer.py (hunk 4 above)
Comment (shtripat) [inline tendrl/node_agent/provisioner/ceph/plugins/ceph_installer.py]: Done

File: tendrl/node_agent/provisioner/ceph/plugins/ceph_installer.py (hunk 5 above)
Comment (shtripat) [inline tendrl/node_agent/provisioner/ceph/plugins/ceph_installer.py]: Done

**Comment by shtripat** [comment]
Verified the changes after incorporation of review comments again and it works as expected.

File: tendrl/node_agent/provisioner/ceph/plugins/ceph_installer.py [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+        # a mon for an existing cluster, provide a list of objects representing
+        # the monitor host and its interface or address.
+
+        data = {
+            "calamari": False,
+            "conf": {"global": {"auth supported": "cephx"}},
+            "host": host,
+            "interface": "eth0",
+            "fsid": cluster_id,
+            "monitor_secret": "AQA7P8dWAAAAABAAH/tbiZQn/40Z8pr959UmEA==",
```
Comment (r0h4n) [inline tendrl/node_agent/provisioner/ceph/plugins/ceph_installer.py]: why have you hardcoded the monitor_secret?

Comment (r0h4n) [inline tendrl/node_agent/provisioner/ceph/plugins/ceph_installer.py]: @nthomas-redhat

Comment (shtripat) [inline tendrl/node_agent/provisioner/ceph/plugins/ceph_installer.py]: we can have this configurable if required. earlier also we had this hardcoded...

**Pull request merged**